docs/proposals/04-etcd-member-custom-resource.md

```diff
@@ -0,0 +1,545 @@
+
+* Etcd-cluster member list.  to check if it is already part of the cluster. 
+
+* Existing etcd data directory and its validity.
+
+> **NOTE:** When the etcd-member gets promoted to a voting-member, then it should remove the annotation on its respective `EtcdMember` resource.
+
+### TLS Enablement for Peer Communication
+
+Etcd-members in a cluster use [peer URL(s)](https://etcd.io/docs/v3.4/op-guide/configuration/#--initial-advertise-peer-urls) to communicate amongst each other. If the advertised peer URL(s) for an etcd-member are updated then [etcd mandates a restart](https://etcd.io/docs/v3.4/op-guide/runtime-configuration/#update-advertise-peer-urls) of the etcd-member.
+
+Druid only supports toggling the transport level security for the advertised peer URL(s). To indicate that the etcd process within the etcd-member has the updated advertised peer URL(s), an annotation `member.etcd.gardener.cloud/tls-enabled` is added by backup-sidecar container to the member lease object.
+
+During the reconciliation run for an `Etcd` resource in druid, if reconciler detects a change in advertised peer URL(s) TLS configuration then it will watch for the above mentioned annotation on the member lease. If the annotation has a value of `false` then it will trigger a restart of the etcd-member pod. 
+
+The authors propose to publish member metadata information in `EtcdMember` resource and not misuse member [lease](https://kubernetes.io/docs/concepts/architecture/leases/)s.
+
+```yaml
+peerTLSEnabled: <bool>
+```
+
+### Monitoring Backup Health
+
+Backup-sidecar takes delta and full snapshot both periodically and threshold based. These backed-up snapshots are essential for restoration operations for bootstrapping an etcd cluster from 0 -> 1 replicas. It is essential that leading-backup-sidecar container which is responsible for taking delta/full snapshots and uploading these snapshots to the configured backup store, publishes this information for druid to consume.
+
+At present, information about backed-up snapshot (*only* `latest-revision-number`) is published by leading-backup-sidecar container by updating `Spec.HolderIdentity` of the delta-snapshot and full-snapshot [lease](https://kubernetes.io/docs/concepts/architecture/leases/)s.
+
+Druid maintains `conditions` in the `Etcd` resource status, which include but are not limited to maintaining information on whether backups being taken for an etcd cluster are healthy (up-to-date) or stale (outdated in context to a configured schedule). Druid computes these conditions using information from full/delta snapshot leases.
+
+In order to provide a holistic view of the health of backups to human operators, druid requires additional information about the snapshots that are being backed-up. The authors propose to not misuse [lease](https://kubernetes.io/docs/concepts/architecture/leases/)s and instead publish the following snapshot information as part `EtcdMember` custom resource:
+
+```yaml
+snapshots:
```

Comment:
IIRC, only the leader backs up the data and since a leader change over time is possible, different `EtcdMember` resources have different status information about snapshots.

Reply:
That is right, only the leading backup-sidecar will take snapshots, however since it is possible that due to re-election there is a leadership change any EtcdMember whose backup-sidecar is a leader can then start to record this information. Since all this information has a timestamp associated to it, it helps analyse the following:

- Druid can check the leader and inspect the published snapshot meta-info to find if its been published regularly.
- Since all EtcdMember resources will have this, you can get a holistic and near-historic information about earlier snapshots taken by other EtcdMember's. This is only meant for diagnostic.

The structure of `EtcdMember` will be kept consistent and we will keep a limit to the number of snapshot information that we capture. This is not yet decided.